src/app/util/attribute-storage.cpp

```diff
@@ -1451,6 +1423,64 @@ app::AttributeAccessInterface * GetAttributeAccessOverride(EndpointId endpointId
 
     return nullptr;
 }
+
+EmberAfStatus SetParentEndpointForEndpoint(EndpointId childEndpoint, EndpointId parentEndpoint)
+{
+    uint16_t childIndex  = emberAfIndexFromEndpoint(childEndpoint);
+    uint16_t parentIndex = emberAfIndexFromEndpoint(parentEndpoint);
+
+    if (childIndex == kEmberInvalidEndpointIndex || parentIndex == kEmberInvalidEndpointIndex)
+    {
+        return EMBER_ZCL_STATUS_UNSUPPORTED_ENDPOINT;
+    }
+    emAfEndpoints[childIndex].parentEndpointId = parentEndpoint;
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+EmberAfStatus SetFlatCompositionForEndpoint(EndpointId endpoint)
+{
+    uint16_t index = emberAfIndexFromEndpoint(endpoint);
+    if (index == kEmberInvalidEndpointIndex)
+    {
+        return EMBER_ZCL_STATUS_UNSUPPORTED_ENDPOINT;
+    }
+    emAfEndpoints[index].bitmask.Clear(EmberAfEndpointOptions::isTreeComposition);
+    emAfEndpoints[index].bitmask.Set(EmberAfEndpointOptions::isFlatComposition);
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+EmberAfStatus SetTreeCompositionForEndpoint(EndpointId endpoint)
+{
+    uint16_t index = emberAfIndexFromEndpoint(endpoint);
+    if (index == kEmberInvalidEndpointIndex)
+    {
+        return EMBER_ZCL_STATUS_UNSUPPORTED_ENDPOINT;
+    }
+    emAfEndpoints[index].bitmask.Clear(EmberAfEndpointOptions::isFlatComposition);
+    emAfEndpoints[index].bitmask.Set(EmberAfEndpointOptions::isTreeComposition);
```

Comment:
Since it's not possible to have both, could we have just `isFlatComposition` being determined via `!isTreeComposition` ?

Reply:
I might leave both functions because you can get a `false` if you provide an invalid endpoint:

```
if (index == kEmberInvalidEndpointIndex)
    {
        return false;
    }
```